KIALI-2529 Update from react-scripts-ts to React-scripts

cfcosta opened the pull request:
## Motivation

There are two motivations for this change:

* `react-scripts-ts` has been deprecated now what the original `react-scripts`
  now [supports typescript natively][1].
* `webpack-dev-server` is [vulnerable for development][2]. It doesn't seem to
  affect our use case, but it throws false positives on the vulnerability
  ratings we do have (`snyk`, `yarn audit`).

Because of that, it is something that we have to do, that will become harder as
we need for a perfect time to do the upgrade, but that isn't urgent enough that
we can't move it forward for after 1.0.

## Strategy

We did a huge PR in the beginning that included all the changes that we had to
do, but that revealed itself to be quite unwieldly. Long PRs require a lot of
context, can hide lots of small errors and is generally hard to review, so
people just don't do it.

To avoid that, we decided to go with two different "sets" of pull requests, one
for "code changes" and the other for library changes. Library changes are
updates on library versions, and are more risky because we don't control the
code that's being changed, while the code changes are usually things that did
compile before but does not anymore, mostly because the compilation is now a
little different than it was before, because `react-scripts` uses Babel for
transpiling, and as such has some differences from the old implementation
(namely namespaces).

The idea is that all the pull requests are independent from each other, and can
be merge at a more suitable time for each of them. Code changes are supposed to
keep working on the current version while also preparing the codebase for the
library changes.

## Dependencies

- [x] #1225 Remove deprecated `namespace`

  **Important**: This PR is not as huge as it seems, add `?w=1` to the URL to
  see the changes without the whitespace changes.

- [x] #1123 Remove deprecated `const enum`
- [x] #1124 Fix cases of type erasure in the codebase that don't compile anymore
- [x] #1236 Fix absolute imports
- [x] #1237 Fix moment import
- [x] #1238 Fix export problems
- [x] #1239 Fix values that can be undefined or null
- [x] #1240 Add type hints to composite types
- [x] #1241 Fix compilation error (this one is a treat, go read it)
- [x] #1222 Update to latest version of `typescript`
- [x] #1263 Fix compilation error still present after #1240 
- [x] Fix failing tests

[1]: https://vincenttunru.com/migrate-create-react-app-typescript-to-create-react-app/
[2]: https://www.npmjs.com/advisories/725

jotak commented:
@cfcosta I've also some troubles to use react-scripts in k-charted. The issue I have is react-script is overriding the tsconfig file hence we cannot configure TS as we would like.
I guess this is the same here.
At the moment I ended up using jest directly, and not through react-script, in order to run tests and have CI pass.

jotak commented:
(see https://github.com/kiali/k-charted/pull/4 )

aljesusg commented:
Why we are moving to react-scripts?

jotak commented:
@aljesusg the *-ts is deprecated as react-script now handles typescript natively: https://github.com/wmonk/create-react-app-typescript#-deprecated-

aljesusg commented:
> @aljesusg the *-ts is deprecated as react-script now handles typescript natively: https://github.com/wmonk/create-react-app-typescript#-deprecated-

Ok Understood @jotak , I asked because I read the JIRA and there was not definition about this. Where can I see the security warnings that the last version of PF @cfcosta ?

jotak commented:
@cfcosta I'm adding a new label here "post-GA", thinking it shouldn't be merged in master. I fear it could bring some instability while we're stabilizing / fixing issues mostly. Does it sound good?

cfcosta commented:
Did an update to the description, this PR is the reference for the changes on this, and it depends on many others.

cfcosta commented:
@mtho11 https://travis-ci.org/kiali/kiali-ui/builds/547495892?utm_source=github_status&utm_medium=notification I just saw that on Travis CI warnings are being treated as errors, and there are many warnings, no way we can tackle all at once.

mtho11 commented:
@cfcosta Is that because of this:
```
WARNING: You are currently running a version of TypeScript which is not officially supported by typescript-estree.
You may find that it works just fine, or you may not.
SUPPORTED TYPESCRIPT VERSIONS: >=3.2.1 <3.5.0
YOUR TYPESCRIPT VERSION: 3.5.1
```

cfcosta commented:
@mtho11 nope, that's because of the compilation warnings that happen later. This is not being treated as a compilation error.

mtho11 commented on package.json:
```diff
@@ -127,7 +126,9 @@
     "redux-mock-store": "1.5.3",
     "snyk": "^1.161.1",
     "source-map-explorer": "1.8.0",
+    "tslint": "^5.17.0",
```
Can we get rid of the `^` and pin it to specific versions?
Snyk is the only one that should have `^`.
`tslint-react` also has `^`

mtho11 commented:
Just FYI, the snapshot tests are failing as well.

mtho11 commented:
The `>10%' fixes android failures.
```
"browserslist": [
    ">10%",
    "last 2 versions",
    "not ie <= 11"
  ],
```

cfcosta commented:
@mtho11 I added it on ba8a9d8 but the errors keep happening, any idea why?

cfcosta commented:
```
Test Suites: 58 passed, 58 total
Tests:       273 passed, 273 total
Snapshots:   31 passed, 31 total
Time:        9.76s
```

Tests are passing, now the only thing left are some warnings, and we can merge this without ignoring warnings on Travis:

```
./src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx
  Line 107:  Anchors must have content and the content must be accessible by a screen reader  jsx-a11y/anchor-has-content

./src/components/MessageCenter/NotificationDrawer.tsx
  Line 108:  The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  jsx-a11y/anchor-is-valid

./src/components/Nav/Menu.tsx
  Line 13:  Using target="_blank" without rel="noopener noreferrer" is a security risk: see https://mathiasbynens.github.io/rel-noopener  react/jsx-no-target-blank

./src/components/IstioWizards/Slider/Slider.tsx
  Line 184:  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text
  Line 186:  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  jsx-a11y/alt-text

./src/components/Filters/StatefulFilters.tsx
  Line 247:  The href attribute requires a valid value to be accessible. Provide a valid, navigable address as the href value. If you cannot provide a valid href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  jsx-a11y/anchor-is-valid

./src/components/Label/Labels.tsx
  Line 52:  The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  jsx-a11y/anchor-is-valid

./src/app/InitializingScreen.tsx
  Line 78:  The href attribute requires a valid value to be accessible. Provide a valid, navigable address as the href value. If you cannot provide a valid href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  jsx-a11y/anchor-is-valid
```

mtho11 commented:
I don't know that we really care about these **_jsx-a11y_** ones yet. AFAIK we have not been testing **_a11y_** compliance yet.


```

./src/components/CytoscapeGraph/ContextMenu/NodeContextMenu.tsx
  Line 107:  Anchors must have content and the content must be accessible by a screen reader  
**jsx-a11y/anchor-has-content**

./src/components/MessageCenter/NotificationDrawer.tsx
  Line 108:  The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md 
 **jsx-a11y/anchor-is-valid**

./src/components/Nav/Menu.tsx
  Line 13:  Using target="_blank" without rel="noopener noreferrer" is a security risk: see https://mathiasbynens.github.io/rel-noopener  
**react/jsx-no-target-blank**

./src/components/IstioWizards/Slider/Slider.tsx
  Line 184:  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images 
 jsx-a11y/alt-text
  Line 186:  img elements must have an alt prop, either with meaningful text, or an empty string for decorative images  
**jsx-a11y/alt-text**

./src/components/Filters/StatefulFilters.tsx
  Line 247:  The href attribute requires a valid value to be accessible. Provide a valid, navigable address as the href value. If you cannot provide a valid href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md  
**jsx-a11y/anchor-is-valid**

./src/components/Label/Labels.tsx
  Line 52:  The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md 
 **jsx-a11y/anchor-is-valid**

./src/app/InitializingScreen.tsx
  Line 78:  The href attribute requires a valid value to be accessible. Provide a valid, navigable address as the href value. If you cannot provide a valid href, but still need the element to resemble a link, use a button and change it with appropriate styles. Learn more: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/anchor-is-valid.md 
 **jsx-a11y/anchor-is-valid**
```

mtho11 commented:
We can probably just  add ` // eslint-disable-next-line` to the line before and deal with them later.

jotak commented on src/config/Logos.ts:
```diff
@@ -1,14 +1,22 @@
-  Go: () => require('../assets/img/go-logo.png'),
-  MicroProfile: () => require('../assets/img/microprofile-logo.png'),
-  JVM: () => require('../assets/img/java-logo.png')
+  'Vert.x': VertxLogo,
```
in the previous version the assets were lazy-loaded, can we get the same with the import statements?

cfcosta replied:
AFAIK it just returns the URL, it does not load the image.

jotak commented on src/pages/Overview/OverviewToolbar.tsx:
```diff
@@ -111,10 +111,17 @@ export class OverviewToolbar extends React.Component<Props, State> {
-    HistoryManager.setParam(URLParam.OVERVIEW_TYPE, otype);
-    this.setState({ overviewType: otype });
-    this.props.onRefresh();
+  updateOverviewType = (otype: String) => {
```
can't we enforce type at compile time as before?

cfcosta replied:
We can't because `handleSelect` takes `(_: String) => any` :(
We can either cast it to the type and ignore the error, or do more complete error handling (which is what I decided to do).

jotak replied:
Ok, I see. Maybe we could make `ToolbarDropdown` use generics, but let's keep that for later :)

cfcosta commented:
@jotak @mtho11 so I tried to find how to tune the linter to ignore the a11y warnings, and realized some things:

* tslint is not used anymore by create-react-app, so we can remove some dependencies
* create-react-app does not support having your own `.eslintrc` on the build script

So I did some patching and black magic and moved to eslint, and now the lint is integrated correctly, correctly ignores a11y errors and found some more warnings that were not fixed before (go figure).

Fixed those, now waiting for Travis to see how much Travis likes my changes, and all the new warnings that appeared were supposed to be fixed before but were not being recognized, for some reason.

josejulio commented:
The bug appears when you go to a workload and navigate to Outbound Metrics tab.
![bug-series-out-metrics](https://user-images.githubusercontent.com/3845764/60205162-188b5e00-9816-11e9-9d44-043e740a29c7.png)

I'll create a JIRA (after lunch :)) with more details, but after a quick debug it looks like we are using k-charted "Histogram" for the runtime and Kiali "Histogram" for type checking

cfcosta commented:
@mattmahoneyrh hey, can you (or someone else on the QE team) run a QE  test on this to make sure we're not breaking anything else?

israel-hdez commented:
The code looks good, but I don't like that black magic :(
I hope we can remove that hack later.

mtho11 commented on package.json:
```diff
@@ -30,33 +30,36 @@
     "prettier": "prettier --write \"{src/**/*.{js,jsx,ts,tsx,json,yml,css,scss},travis.yml,*.json}\"",
-    "snyk-protect": "snyk protect"
+    "snyk-protect": "snyk protect",
+    "postinstall": "sed -i -E 's#\\[.*eslint-config-react-app.*\\]#['\\'$PWD/.eslintrc\\'']#g' node_modules/react-scripts/config/webpack*.js"
```
I don't totally understand this little ditty...

cfcosta replied:
It changes from the default path for `react-scripts`, to stop using the default `eslint` configuration inside `node_modules` to use `.eslintrc` on the project home every time you install the dependencies.

cfcosta replied:
@mtho11 did you test installing the dependencies on os x? I just remembered `sed` has different usage on mac, can you check if it works correctly?

If it doesn't fail and you can run `yarn lint` without any errors then everything is ok :)

jshaughn replied:
Because JSON can't really be commented, I would recommend that this "black magic" explanation be recorded in the final commit comment.  In general, I think we lack explanations for some of the tricky code we leave behind.  There is nothing wrong with helping a future dev with inline or in-commit comments.  Info left in a PR is not as easily traceable.

cfcosta replied:
@jshaughn check this out: https://github.com/kiali/kiali-ui/pull/1154/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R35

But I can add it to the commit message with a more detailed explanation if you think that's not enough.

mtho11 replied:
@cfcosta works fine on OS/X.

mtho11 commented:
We still need to get QE to test though: @mattmahoneyrh

lucasponce commented:
Please @hhovsepy, could you also run a regression test on this ?
Also good if @aljesusg @jotak @xeviknal can take a look as well.

I plan to fetch this PR today and review code locally.

hhovsepy commented:
regression UI tests pass

cfcosta commented:
@lucasponce @aljesusg @xeviknal if you need any help just ping me on this.

I think @jotak already finished testing everything though.

lucasponce commented:
Some problems in the console:
![image](https://user-images.githubusercontent.com/1662329/60284309-85cceb00-990b-11e9-9374-4235762981c0.png)

Not big blockers. I had some JavaScript strange error but I couldn't reproduce it.
@xeviknal @jotak @aljesusg some comments ?

Can you give it a try ?

lucasponce commented:
@jotak also I wonder if now that we have k-charts splitted in a separated library, if we need some extra steps before to merge this.
As I guess some of this work would affect also to k-charts, isn't it ?

lucasponce commented:
I'd prefer to not release if we have a high number of errors in the console:
![image](https://user-images.githubusercontent.com/1662329/60285173-a5fda980-990d-11e9-8211-09c293b8c8b9.png)
Perhaps we can merge after the release, and we can address them in Sprint 25 if some of them required more work.

jshaughn commented:
+1, we should not release with any console errors, if possible.

cfcosta commented:
@jshaughn @lucasponce I'm taking a look at the console errors, and I'm not planning on merging this before the release.

cfcosta commented:
I fixed the errors on the console here, can someone test to help?

mtho11 commented:
@cfcosta I can test this.

mtho11 commented:
@cfcosta @lucasponce I have tested this and can confirm the JS console errors have gone away.

cfcosta commented:
@mtho11 what do you think, are we good to merge?

mtho11 commented:
@cfcosta yep, works for me.

cfcosta commented:
Ok, will do that by the end of the day to give people some time to manifest if they do find anything.

@josejulio @israel-hdez what do you guys think? good to go?

josejulio commented on src/components/CytoscapeGraph/FocusAnimation.ts:
```diff
@@ -59,17 +59,19 @@ export default class FocusAnimation {
       }
 
+      this.clear();
+      this.layer.setTransform(this.context);
```
This code is no longer needed.
The error was caused by running this method without properly binding "this".
We probably want to know when this breaks and this will silently fail.

https://github.com/kiali/kiali-ui/pull/1293

cfcosta replied:
The whole file? This part of the patch? Or the whole commit?

josejulio replied:
Just this file

israel-hdez commented:
@cfcosta For me, it's working OK. Good to merge once you have @josejulio approval :)
